core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import java.util.Locale
+
+import org.apache.spark.sql.delta.actions._
+import org.apache.spark.sql.delta.util.{Utils => DeltaUtils}
+
+import org.apache.spark.sql.SparkSession
+
+object TableFeatureStore {
+
+  /**
+   * All table features recognized by this client. Update this set when you added a new Table
+   * Feature.
+   */
+  val allSupportedFeaturesMap: Map[String, TableFeature] = {
+    var features = Set[TableFeature](AppendOnlyTableFeature)
+    if (DeltaUtils.isTesting) {
+      features ++= Set(
+        TestLegacyWriterFeature,
+        TestWriterFeature,
+        TestLegacyReaderWriterFeature,
+        TestReaderWriterFeature)
+    }
+    val featureMap = features.map(f => f.name.toLowerCase(Locale.ROOT) -> f).toMap
+    require(features.size == featureMap.size, "Lowercase feature names must not duplicate.")
+    featureMap
+  }
+}
+
+/* ---------------------------------------- *
```

Comment:
It is actually more common practice to put the base trait/class first and then implementations. See actions.scala or DeltaOperations.scala.

Reply:
Done! Now it's base class/trait, then the companion object, then features.